### PR TITLE
[HttpKernel] remove all deprecated code from http kernel

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | none   
| License       | MIT
| Doc PR        |to check

REmove all deprecated code from HttpKernel, update all listener and all tests

### CODE REVIEW

**Comment by Simperfit** [comment]
@nicolas-grekas tests are failling because I changed the usage directly in this PR maybe I should do 2 PR ? removed the event  and then change the usage ?

File: src/Symfony/Bridge/Monolog/Processor/RouteProcessor.php [hunk 1]
```diff
@@ -47,7 +47,7 @@ public function reset()
         $this->routeData = [];
     }
 
-    public function addRouteData(GetResponseEvent $event)
+    public function addRouteData(RequestEvent $event)
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/Monolog/Processor/RouteProcessor.php]: that's a BC break: the class should be made final in 4.4

File: src/Symfony/Bundle/FrameworkBundle/EventListener/ResolveControllerNameSubscriber.php [hunk 2]
```diff
@@ -32,7 +32,7 @@ public function __construct(ControllerNameParser $parser)
         $this->parser = $parser;
     }
 
-    public function onKernelRequest(GetResponseEvent $event)
+    public function onKernelRequest(RequestEvent $event)
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/EventListener/ResolveControllerNameSubscriber.php]: the class is deprecated it should be removed

File: src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php [hunk 3]
```diff
@@ -26,21 +26,10 @@ class ConfigDataCollector extends DataCollector implements LateDataCollectorInte
      * @var KernelInterface
      */
     private $kernel;
-    private $name;
-    private $version;
     private $hasVarDumper;
 
-    public function __construct(string $name = null, string $version = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php]: this has already been cleaned up, please rebase

File: src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php [hunk 4]
```diff
@@ -23,6 +26,27 @@
...
+
+    public function __construct(HttpKernelInterface $kernel, callable $controller, array $arguments, Request $request, ?int $requestType)
+    {
+        parent::__construct($kernel, $controller, $request, $requestType);
+
+        $this->arguments = $arguments;
+    }
+
+    /**
+     * @return array
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php]: let's use a real type instead? could require making either the class or the method final in 4.4

File: src/Symfony/Component/HttpKernel/Event/ControllerEvent.php [hunk 5]
```diff
@@ -22,6 +25,29 @@
...
+    {
+        parent::__construct($kernel, $request, $requestType);
+
+        $this->setController($controller);
+    }
+
+    /**
+     * Returns the current controller.
+     *
+     * @return callable
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/ControllerEvent.php]: let's use a real type instead? could require making either the class or the method final in 4.4

File: src/Symfony/Component/HttpKernel/Event/ExceptionEvent.php [hunk 6]
```diff
@@ -24,6 +27,64 @@
...
+
+        $this->setException($e);
+    }
+
+    /**
+     * Returns the thrown exception.
+     *
+     * @return \Exception The thrown exception
+     */
+    public function getException()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/ExceptionEvent.php]: let's use a real type instead? could require making either the class or the method final in 4.4

File: src/Symfony/Component/HttpKernel/Event/RequestEvent.php [hunk 7]
```diff
@@ -20,6 +22,37 @@
...
+    {
+        $this->response = $response;
+
+        $this->stopPropagation();
+    }
+
+    /**
+     * Returns whether a response was set.
+     *
+     * @return bool Whether a response was set
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/RequestEvent.php]: let's use a real type instead? could require making either the class or the method final in 4.4

File: src/Symfony/Component/HttpKernel/Event/ResponseEvent.php [hunk 8]
```diff
@@ -20,6 +24,32 @@
...
+
+        $this->setResponse($response);
+    }
+
+    /**
+     * Returns the current response object.
+     *
+     * @return Response
+     */
+    public function getResponse()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/ResponseEvent.php]: let's use a real type instead? could require making either the class or the method final in 4.4

File: src/Symfony/Component/HttpKernel/Event/ResponseEvent.php [hunk 9]
```diff
@@ -20,6 +24,32 @@
...
+    {
+        return $this->response;
+    }
+
+    /**
+     * Sets a new response object.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/ResponseEvent.php]: let's all these useless docblocks

File: src/Symfony/Component/HttpKernel/Event/TerminateEvent.php [hunk 10]
```diff
@@ -19,6 +23,24 @@
...
+    {
+        parent::__construct($kernel, $request, HttpKernelInterface::MASTER_REQUEST);
+
+        $this->response = $response;
+    }
+
+    /**
+     * Returns the response for which this event was thrown.
+     *
+     * @return Response
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Event/TerminateEvent.php]: let's use a real type instead? could require making either the class or the method final in 4.4

File: src/Symfony/Component/HttpKernel/EventListener/SaveSessionListener.php [hunk 11]
```diff
@@ -24,7 +24,7 @@
  */
 class SaveSessionListener implements EventSubscriberInterface
 {
-    public function onKernelResponse(FilterResponseEvent $event)
+    public function onKernelResponse(ResponseEvent $event)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/EventListener/SaveSessionListener.php]: class is deprecated, let's remove it instead

File: src/Symfony/Component/HttpKernel/KernelInterface.php [hunk 12]
```diff
@@ -122,13 +113,12 @@ public function getEnvironment();
...
-     * Gets the application root dir (path of the project's Kernel class).
+     * Gets the application project dir (path of the project's Kernel class).
      *
-     * @return string The Kernel root dir
+     * @return string The Kernel project dir
      *
-     * @deprecated since Symfony 4.2
      */
-    public function getRootDir();
+    public function getProjectDir();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/KernelInterface.php]: please remove the corresponding `@method` annotation

File: src/Symfony/Component/HttpKernel/Tests/KernelTest.php [hunk 13]
```diff
@@ -82,14 +82,15 @@ public function testInitializeContainerClearsOldContainers()
     {
         $fs = new Filesystem();
         $legacyContainerDir = __DIR__.'/Fixtures/var/cache/custom/ContainerA123456';
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/KernelTest.php]: unrelated

File: src/Symfony/Component/Security/Http/Firewall.php [hunk 14]
```diff
@@ -132,7 +132,7 @@ protected function callListeners(RequestEvent $event, iterable $listeners)
     /**
      * @deprecated since Symfony 4.3, use callListeners instead
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall.php]: hello deprecated method

File: src/Symfony/Component/Security/Http/Firewall/LegacyListenerTrait.php [hunk 15]
```diff
@@ -24,15 +23,15 @@ trait LegacyListenerTrait
     /**
      * @deprecated since Symfony 4.3, use __invoke() instead
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/LegacyListenerTrait.php]: hello deprecation (the class itself is deprecated)

File: src/Symfony/Component/Security/Http/Firewall/ListenerInterface.php [hunk 16]
```diff
@@ -22,5 +22,5 @@
  */
 interface ListenerInterface
 {
-    public function handle(GetResponseEvent $event);
+    public function handle(RequestEvent $event);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/ListenerInterface.php]: this interface is deprecated

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Please also turn all `@final since Symfony X.Y` to just `@final`,
same for `@internal etc`
PRs must be green yes. If splitting them makes it easier, then that's the way.
Please be very careful when doing removal PRs, there are many things to take care of (eg see attached review)

File: src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/Event/ControllerArgumentsEvent.php]: oh wait sorry I forgot about #30323 which will allow this everywhere!

File: src/Symfony/Component/Security/Http/Firewall.php (hunk 14 above)
Comment (Simperfit) [reply src/Symfony/Component/Security/Http/Firewall.php]: going to create a PR for this.

File: src/Symfony/Component/Security/Http/Firewall/LegacyListenerTrait.php (hunk 15 above)
Comment (Simperfit) [reply src/Symfony/Component/Security/Http/Firewall/LegacyListenerTrait.php]: going to create a PR for this.

File: src/Symfony/Component/Security/Http/Firewall/ListenerInterface.php (hunk 16 above)
Comment (Simperfit) [reply src/Symfony/Component/Security/Http/Firewall/ListenerInterface.php]: going to create a PR for this.

File: src/Symfony/Component/Security/Http/Firewall.php (hunk 14 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Http/Firewall.php]: See #31782 for the security part. Once merged, you should be able to revert all changes from the `Security\Http` namespace here.

**Comment by Simperfit** [comment]
PR Rebased

File: UPGRADE-5.0.md [hunk 17]
```diff
@@ -252,7 +252,7 @@ HttpKernel
 ----------
 
  * Removed `Client`, use `HttpKernelBrowser` instead
- * The `Kernel::getRootDir()` and the `kernel.root_dir` parameter have been removed
+ * The `Kernel::getRootDir()` and the `kernel.root_dir` parameter have been removed use `Kernel::getProjectDir()` and the `kernel.project_dir` instead
```
Comment (fabpot) [inline UPGRADE-5.0.md]: `removed, use` (comma). Also, can you cut at approx 80 chars?

File: src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md [hunk 18]
```diff
@@ -9,6 +9,7 @@ CHANGELOG
  * Removed `ResolveControllerNameSubscriber`
  * Removed support for `bundle:controller:action` to reference controllers. Use `serviceOrFqcn::method` instead
  * Removed support for PHP templating, use Twig instead
+ * Remove `ResolveControllerNameSubscriber`
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/CHANGELOG.md]: Removed

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml [hunk 19]
```diff
@@ -87,9 +87,9 @@
 
         <service id="file_locator" class="Symfony\Component\HttpKernel\Config\FileLocator">
             <argument type="service" id="kernel" />
-            <argument>%kernel.root_dir%/Resources</argument>
+            <argument>%kernel.project_dir%/Resources</argument>
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: That looks suspicious: `root_dir` and `project_dir` are not equivalent.

File: src/Symfony/Component/HttpKernel/CHANGELOG.md [hunk 20]
```diff
@@ -8,6 +8,21 @@ CHANGELOG
  * removed `ConfigDataCollector::getApplicationName()` 
  * removed `ConfigDataCollector::getApplicationVersion()`
  * removed support for `Symfony\Component\Templating\EngineInterface` in `HIncludeFragmentRenderer`, use a `Twig\Environment` only
+ * removed `Client`, use `HttpKernelBrowser` instead
+ * the `Kernel::getRootDir()` and the `kernel.root_dir` parameter have been removed
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/CHANGELOG.md]: removed the ... (same below)

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (hunk 19 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: What will be the equivalent since this is to me the only thing I can think of to replace root_dir here.

File: src/Symfony/Bridge/Monolog/composer.json [hunk 21]
```diff
@@ -19,7 +19,7 @@
         "php": "^7.2.9",
         "monolog/monolog": "~1.19",
         "symfony/service-contracts": "^1.1",
-        "symfony/http-kernel": "^4.4|^5.0"
+        "symfony/http-kernel": "^5.0"
```
Comment (Tobion) [inline src/Symfony/Bridge/Monolog/composer.json]: for my information, why is this necessary? shouldn't using the  new event classes also work in 4.4?

Comment (Simperfit) [reply src/Symfony/Bridge/Monolog/composer.json]: I don't think this is necessary will revert

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (hunk 19 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: I had to look in the code what this does: This is the global directory where resources can be loaded from. `%kernel.project_dir%/Resources` and `%kernel.project_dir%` from below are the fallback directories for any resources. So in SF 3 this means
- `app/Resources` can be used to overwrite resources from Bundles
- `app/` can be used to load any resource if they are not found relative to the current directory the source is loaded from

But that behavior has never officially been unsed in a SF 2/3 symfony-standard or SF 4 flex kernel. They are either loaded relative to the current dir or via glob pattern, not by a global fallback.
Overwriting bundle resources should also not be done like this since forever. See https://symfony.com/doc/current/bundles/override.html
It should be done with compiler passes for services or twig namespace (which uses the /templates/ dir) for templates or translation domains for translations.

If we change this to use project_dir, it means the fallback dir is not `app/Resources` as in SF 3 or `src/Resources` in SF 4, but `Resources/` directly in SF 5. Whatever we put here, it does not make sense and is not documented anywhere.

I suggest to deprecate this global fallback dir in SF 4.4 and remove it in 5. Then this problem is gone.
I already have an idea how to implement the deprecation.

File: src/Symfony/Component/HttpKernel/Tests/KernelTest.php [hunk 22]
```diff
@@ -309,36 +309,6 @@ public function doStuff()
...
-
-        $this->assertEquals('Fixtures', $kernel->getName());
-    }
-
-    /**
-     * @group legacy
-     */
-    public function testOverrideGetName()
-    {
-        $kernel = new KernelForOverrideName('test', true);
```
Comment (Tobion) [inline src/Symfony/Component/HttpKernel/Tests/KernelTest.php]: this class should be removed

File: src/Symfony/Bridge/Monolog/composer.json (hunk 21 above)
Comment (Tobion) [reply src/Symfony/Bridge/Monolog/composer.json]: this uneccessary dependency raise happend in several places

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (hunk 19 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: Thanks for looking into it @Tobion If you have time to implement the deprecation go ahead, if oyu want me to do it, tell me what you think is the best way to do it!

**Comment by Simperfit** [comment]
The events test has been fixed, still missing the resolution of https://github.com/symfony/symfony/pull/31672#discussion_r290758431 and the some kernel.root_dir test are broken, this needs to be looked at to maybe raise the dependencies.

**Comment by nicolas-grekas** [comment]
To me, the events part should be split in its own PR.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml (hunk 19 above)
Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: I extracted this issue into #31915. Just merge it like this now and we can deal with this proble separately.

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml]: I reverted the removal of the `kernel.root_dir` parameter from this PR so we can work on #31915 before.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @Simperfit.
